### PR TITLE
explicit opts.nthreads control

### PR DESCRIPTION
**Pull request by ahbarnett**
The logic is opts.nthreads=0 by default, which causes finufft.cpp to query omp_get_max_threads
(which is set by OMP_NUM_THREADS env, and by in various languages) and resets opts.nthreads to that. Alternatively, opts.nthreads>0 overrides the value, without upper limit (in case users want to try crazy large).

opts.nthreads is then used throughout to set omp parallel blocks via explicit num_threads() clauses. (The sorter is different and can use 1 or this number).
The beauty is, the omp state is never changed, so we don't have to think about the horror of restoring it (I could find very little out about doing this anyway in omp docs).

This works from C++, C, Fortran, Matlab, octave.   It needs to be brought into python.

### CODE REVIEW

File: python/finufftpy/_interfaces.py [hunk 1]
```diff
@@ -22,9 +22,16 @@
...
+### Plan class definition
+class Plan:
+    def __init__(self,tp,n_modes_or_dim,iflag=None,n_trans=1,eps=None,**kwargs):
+        # set default iflag based on if iflag is None
+        if iflag==None:
+            if tp==2:
+                iflag = -1
+            else:
+                iflag = 1
+
```
Comment (janden) [inline python/finufftpy/_interfaces.py:34]: (This is for below.) Didn't we say that default epsilon should be `1e-6` regardless of precision?

File: src/finufft.cpp [hunk 2]
```diff
@@ -195,25 +196,23 @@ void onedim_fseries_kernel(BIGINT nf, FLT *fwkerhalf, spread_opts opts)
     a[n] = exp(2*PI*IMA*(FLT)(nf/2-z[n])/(FLT)nf);  // phase winding rates
   }
   BIGINT nout=nf/2+1;                   // how many values we're writing to
-  int nt = min(nout,(BIGINT)MY_OMP_GET_MAX_THREADS());      // how many chunks
+  int nt = min(nout,(BIGINT)opts.nthreads);         // how many chunks
   std::vector<BIGINT> brk(nt+1);        // start indices for each thread
   for (int t=0; t<=nt; ++t)             // split nout mode indices btw threads
     brk[t] = (BIGINT)(0.5 + nout*t/(double)nt);
-#pragma omp parallel
+#pragma omp parallel num_threads(nt)
```
Comment (janden) [inline src/finufft.cpp:203]: This is great.

File: test/finufftGuru_test.cpp [hunk 3]
```diff
@@ -171,36 +171,28 @@ int main(int argc, char* argv[])
...
     printf("ntr=%d: %lld NU pts to %lld modes in %.3g s \t%.3g NU pts/s\n", ntransf, (long long)M,(long long)N, totalTime, ntransf*M/totalTime);
   else
     printf("ntr=%d: %lld NU pts to %lld NU pts in %.3g s \t%.3g tot NU pts/s\n", ntransf, (long long)M,(long long)N, totalTime, ntransf*(N+M)/totalTime);
 
-  
   // Comparing timing results with repeated calls to corresponding finufft function...
-  FFTW_CLEANUP();
-  FFTW_CLEANUP_THREADS();
-  FFTW_FORGET_WISDOM();
+  //FFTW_CLEANUP();
```
Comment (janden) [inline test/finufftGuru_test.cpp:181]: Remove? Any need to keep commented-out code?

File: python/finufftpy/_interfaces.py (hunk 1 above)
Comment (lu1and10) [reply python/finufftpy/_interfaces.py:34]: I thought we agreed with your suggestion, in the email you said the following,

> * tol. My concern with leaving it at 1e-6 (even for double inputs) is that new users may assume Finufft is simply too inaccurate for their purposes (if they thought 1e-12 was inaccurate enough, imagine how horrified they would be by 1e-6). A safer default might then be to have tolerance being as high as possible for the input datatype (so 1e-6 for single precision, 1e-14 for double precision). We routinely run the 3D code at this precision in Aspire and aren't too concerned about the performance (although maybe we should be?). I think the competition essentially fixes their tolerance at 1e-12 (for whatever that's worth).

I guess I might miss the final discussion for default eps in the Zoom meeting?

Comment (janden) [reply python/finufftpy/_interfaces.py:34]: Right, but I think we discussed it in the meeting and we ended up deciding on a fixed tolerance of `1e-6` for all precision. @ahbarnett can you confirm?

Comment (ahbarnett) [reply python/finufftpy/_interfaces.py:34]: I confirm that we decided 1e-6 is default for both precisions. (It's unrealistic for 1d single-prec xforms, but, whatever..)

File: src/finufft.cpp (hunk 2 above)
Comment (ahbarnett) [reply src/finufft.cpp:203]: You mean, you like explicit num_threads throughout? So do I :)

BTW, I did ad hoc speed comparisons on the i7/GCC of schedule(static) vs schedule(dynamic,1) vs dynamic,10000, etc.   There isn't much logic to it, but explains some of my changes.

File: test/finufftGuru_test.cpp (hunk 3 above)
Comment (ahbarnett) [reply test/finufftGuru_test.cpp:181]: this test code is a bit messy, but persists as a general command-line guru driver. These FFTW commands are left to remind us that Andrea found they were needed to get correct timings for the next finufft call.

File: src/finufft.cpp (hunk 2 above)
Comment (janden) [reply src/finufft.cpp:203]: Yes. This feels very much like the Right Thing to do…

File: test/finufftGuru_test.cpp (hunk 3 above)
Comment (janden) [reply test/finufftGuru_test.cpp:181]: All right. So can we uncomment them? If not, maybe leave an explanatory comment about timings?

File: python/test/guru1d1.py [hunk 4]
```diff
@@ -20,10 +18,10 @@
...
 #plan
-plan = fp.FinufftPlan(1,n_modes,iflag,1,acc,debug=1)
+plan = fp.Plan(1,(N,))
 
 #set pts
-plan.setpts(x,None,None,None,None,None)
+plan.setpts(x)
 
 #exec
 plan.execute(c,F)
```
Comment (janden) [inline python/test/guru1d1.py:27]: Still should have `F = plan.execute(c)` here.

**Comment by ahbarnett** [comment]
I'll tweak then bring in this PR tomorrow am. Got to sleep now :)

File: test/finufftGuru_test.cpp (hunk 3 above)
Comment (ahbarnett) [reply test/finufftGuru_test.cpp:181]: Commented. I declare this no longer part of the main testers, unsupported.

File: examples/guru1d1c.c [hunk 5]
```diff
@@ -39,6 +39,7 @@ int main(int argc, char* argv[])
     popts = (nufft_opts *)malloc(sizeof(nufft_opts));         // allocate it
     finufft_default_opts(popts);
     popts->debug = 1;        // example options change
+    popts->nthreads = 4;     // "
```
[the remark this answers is not in the document]
Comment (ahbarnett) [reply examples/guru1d1c.c:42]: no, the point is to change from the default :)

File: makefile [hunk 6]
```diff
@@ -146,9 +151,9 @@ HEADERS = $(wildcard include/*.h)
 	$(CC) -c $(CFLAGS) $< -o $@
 %_32.o: %.c $(HEADERS)
 	$(CC) -DSINGLE -c $(CFLAGS) $< -o $@
-%.o: %.f $(HEADERS)
+%.o: %.f
```
[the remark this answers is not in the document]
Comment (ahbarnett) [reply makefile:154]: no, fortran doesn't look at headers.

File: src/finufft.cpp (hunk 2 above)
Comment (ahbarnett) [reply src/finufft.cpp:203]: :)

File: test/finufftGuru_test.cpp (hunk 3 above)
Comment (ahbarnett) [reply test/finufftGuru_test.cpp:181]: Commented. I declare this an unsupported test code and will move to perftests which is more for devs.

File: examples/guru1d1c.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (janden) [reply examples/guru1d1c.c:42]: Ah right. Sorry wasn't thinking.

File: makefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (janden) [reply makefile:149]: Got it.

**Pull request merged**